### PR TITLE
Paper 8 neural quantum embedding pushing the limits of quantum supervised learning

### PR DESCRIPTION
**Pull request by LF-Vigneux**
Reproducing  “[Neural Quantum Embedding: Pushing the Limits of Quantum Supervised Learning](https://arxiv.org/abs/2311.11412v2 )” by Hur et al., 2024 .

### CODE REVIEW

File: papers/nn_embedding/README.md [hunk 1]
```diff
@@ -0,0 +1,554 @@
...
+Code provided: https://github.com/takh04/neural-quantum-embedding  
+
+## 1. Executive Summary
+### What the paper does
+
+This paper introduce a QPU efficient way to encode classical data. This methods, proven mathematically, enables the encoding of the data to be maximally distant on the quantum Hilbert space. It also states that the rest of the trainable circuit can not improve that embedding and can just find an optimal measurement limited by the representation of the data.
+
+### Why it matters
+
+An optimal encoding makes the classification of the data easier. As it is showed in the paper, in can severely limit the performance of the model if it is not appropriate. The performance of an embedding varies on the dataset. Hence, a variable encoding may be the optimal way to classify the data. The classifier part of a QML algorithm only finds the optimal measurment that is clearly limited by the encoded states.
```
Comment (CassNot) [inline papers/nn_embedding/README.md]: measurement

File: papers/nn_embedding/README.md [hunk 2]
```diff
@@ -0,0 +1,554 @@
...
+The optimal encoded limits the expressivity of the model, and so the trainability.
+
+It still shows advantages in noisy simulations and on hardware.
+
+### Bottom line 
+> Reproduced (Except one figure)  
+
+### Main takeaways 
+
+The MerLin implementation was difficult but it is efficient
```
Comment (CassNot) [inline papers/nn_embedding/README.md]: A bit unclear: could you specify why it was difficult and why it is efficient ? (I am not sure why these 2 adjectives are in opposition)

File: papers/nn_embedding/README.md [hunk 3]
```diff
@@ -0,0 +1,554 @@
...
+
+The same results are observed with the gate-based version and MerLin-based version
```
Comment (CassNot) [inline papers/nn_embedding/README.md]: For the gate-based, you can precise the language

File: papers/nn_embedding/README.md [hunk 4]
```diff
@@ -0,0 +1,554 @@
...
+- Phase 1: A classical model takes the classical features in input, generates weights for the embedding part of the quantum circuit. We train the classical model to maximise the distance bewteen encoded states of different classes.
+
+- Phase 2: Then we train the classifier part of the quantum circuit just like any QML algorithm. The parameters of the classical models are here frozen, but the input still goes through this model.
+
+### Main figure / pipeline
+
+Phase 1:
+![](img_paper/phase_1_paper.png)
+
+Phase 2:
```
Comment (CassNot) [inline papers/nn_embedding/README.md:56]: Detail: the phase 2 appears below the image

File: papers/nn_embedding/README.md [hunk 5]
```diff
@@ -0,0 +1,554 @@
...
+### Architecture
+| Paper                         | Photonic reimplementation |
+|------------------------------|---------------------------|
+| 4 qubits, PCA 8              | n=2, m=6, PCA 8          |
+| 8 qubits, PCA 8              | n=5, n=10, PCA 8         |
+| 4 qubits, PCA 4              | n=2, m=6, PCA 4          |
+
+**Notes:**  
+The classical models are composed of 3 linear layers separated by the ReLU activation function.  
+For the reimplementation, the quantum embedding takes more parameters than the gate-based model.
```
Comment (CassNot) [inline papers/nn_embedding/README.md]: Is it because of the number of parameters induced in the merlin version ?

File: papers/nn_embedding/README.md [hunk 6]
```diff
@@ -0,0 +1,554 @@
...
+
+---
+
+### Training setup
+- **Paper:** MNIST dataset with digits 0 and 1  
+- **Photonic reimplementation:** Same setup  
+- **Notes:** Always using the ADAM optimiser (the paper uses Nesterov)
```
Comment (CassNot) [inline papers/nn_embedding/README.md]: Why this choice for the optmizer ?
NB: Torch proposed [NAdam](https://docs.pytorch.org/docs/2.11/generated/torch.optim.NAdam.html) which incorporated Nesterov Momentum in Adam
No need to redo the experiments but maybe a statement on what are the main differences and expected behaviour differences in the training could be needed

File: papers/nn_embedding/README.md [hunk 7]
```diff
@@ -0,0 +1,554 @@
...
+- Learning rate: **0.01**  
+- Distance used in training: **trace distance**
+
+---
+
+### Missing details / assumptions
+- The number of epochs per step is not always clarified  
+- It will be specified per figure in section 7.2.
+
+To use later \begin{equation}
```
Comment (CassNot) [inline papers/nn_embedding/README.md]: This does not render correctly 
What does you epsilon it refer to ?

File: papers/nn_embedding/README.md [hunk 8]
```diff
@@ -0,0 +1,554 @@
...
+
+#### Deviations from paper:
+
+- The quantum embedding parameters that are optimized are all the trainable ones in the MerLin QuantumLayer. There is a better control of the trainable parameters in the gate-based version.
+
+## 6. Reproduction results
+
+### Result status
+
+Figures 2 through 6 are reproduced and a noise-free simulation with Pennylane. Some better or equivalent results are observed with the MerLin implementation.
```
Comment (CassNot) [inline papers/nn_embedding/README.md]: I think the first sentence misses a verb wrt noise-free simulation

File: papers/nn_embedding/README.md [hunk 9]
```diff
@@ -0,0 +1,554 @@
...
+
+All of the parameters values used are available in the table of section 7.2.
+
+### 6.1. Figure 2
+
+We want to compare NQE to a fixed embedding classifier.
+
+This identifies the learning process of our NQE model of learning the 0 and 1 digits of the MNIST dataset. We analyze during the embedding training phase the trace distance between the average encoded states of each class. We also plot the loss of the models when training afterwards the classifier. We also indicate the final test accuracies of the model on the loss plot.
+
```
Comment (CassNot) [inline papers/nn_embedding/README.md:165]: Could you add a note on how you implemented NQE with MerLin ?

File: papers/nn_embedding/README.md [hunk 10]
```diff
@@ -0,0 +1,554 @@
...
+We clearly observe the same trends as the paper for the reproductions showing a clear advantage for NQE methods against trainable encodings. Better accuracy across and faster loss decrease is observed just like with fixed embeddings in figure 2. The MerLin advantage versus trainable unitaries seems to be larger.
+
+
+### 6.3. Figure 4
+In QML, there is a tradeoff between expressivity and trainability. Indeed, a larger expressivity is needed for complex data, but, with the barren plateau phenomena, it makes the model difficult to train. With a optimal encoded, we can limit the effective dimension, simplify the problem and make the model easier to train. That is why a study of the effective dimension (a metric also used in classical ML) makes sense here. Here is the formula.
+$$d_{LED}=\frac{2\log\bigg(\frac{1}{V_\epsilon}\int_{B_\epsilon (\theta^*)}\sqrt{\det(\mathbb{1}+\kappa_{n,\gamma}\overline{F}(\theta))}\mathrm{d} \theta\bigg)}{\log (\kappa_{n,\gamma})}$$
+
+For more details about the parameters of the local effective dimension and the way it is computed here. Checkout section 4 in this [LaTeX document](https://www.overleaf.com/project/69a9c53955a12333eab55974).
```
Comment (CassNot) [inline papers/nn_embedding/README.md]: I do not have permissions to open this page
Maybe you can save it as a pdf in utils/ ?

File: papers/nn_embedding/README.md [hunk 11]
```diff
@@ -0,0 +1,554 @@
...
+![](results/figure_6/figure_6_merlin.png)
+
+#### Discussion
+We obtain the same conclusion as the figure from the paper we get a large deviation from the two design with the NQE methods that without it. We even get a better difference with the reproductions and even more in photonics. The formula uses an integral over the Haar distribution. For the photonic implementation, it was replaced by an integral over random states creates from torch. random. Since the spaces are not the same, it is difficult to compare the values directly. We still can conclude that the space explored by the embedding trained model is a smaller space that isn easier to train. We conclude the same thing with the kernel variance plot where it is bigger with NQE.
+
+## 7. Photonic translation
+
+#### Photonic objective
+
+- Obtain the same or bette results than the gate-based model.
```
Comment (CassNot) [inline papers/nn_embedding/README.md]: bette**r**
Can you specify again when it is the same and when it is better ?

File: papers/nn_embedding/README.md [hunk 12]
```diff
@@ -0,0 +1,554 @@
...
+
+#### Circuit / model
+
+- The embedding and classifying circuits are a trainable universal unitary with the add_entangling_layer method of the CircuitBuilder.
+
+### 7.1. MerLin feasibility
+
+#### Can this be done in MerLin? 
+
+- The entire pipeline was reproduced in MerLin and is even faster than the Pennylane implementation.
```
Comment (CassNot) [inline papers/nn_embedding/README.md]: Do you have a factor for the computational speed-up ?

File: papers/nn_embedding/README.md [hunk 13]
```diff
@@ -0,0 +1,554 @@
...
+
+- Obtain the same or bette results than the gate-based model.
+
+#### Proposed photonic formulation
+
+- The encoding part of the interferometer has no input parameters and the classical model changes all of the trainable parameters.
+
+#### Circuit / model
+
+- The embedding and classifying circuits are a trainable universal unitary with the add_entangling_layer method of the CircuitBuilder.
```
Comment (CassNot) [inline papers/nn_embedding/README.md]: classifier instead of classifying ?

File: papers/nn_embedding/configs/defaults.json
Comment (CassNot) [inline papers/nn_embedding/configs/defaults.json:1]: All results folder should be by default named "outdir" 
This comment is valid for all config files

File: papers/nn_embedding/nn_embedding.ipynb
Comment (CassNot) [inline papers/nn_embedding/nn_embedding.ipynb:1]: I would add the main Figure of the paper for contextualization

Comment (CassNot) [reply papers/nn_embedding/nn_embedding.ipynb:1]: Otherwise, very great notebook !
I would maybe display (pcvl.pdisplay) the circuit used in the QuantumLayer

**Review by CassNot — COMMENTED** [review]
Very good work !

- Most of the required changes are related to the README
- 2 changes required in the notebook
- the results folder by default should be named "outdir"

**Review by CassNot — CHANGES_REQUESTED** [review]
Very good work !

- Most of the required changes are related to the README
- 2 changes required in the notebook
- the results folder by default should be named "outdir"

File: papers/nn_embedding/README.md (hunk 10 above)
Comment (LF-Vigneux) [reply papers/nn_embedding/README.md]: Changed the link instead

File: papers/nn_embedding/README.md (hunk 12 above)
Comment (LF-Vigneux) [reply papers/nn_embedding/README.md]: I do not have it, the speed advantage is significant but, to be rigour, I changed the verb from be to seem.

File: papers/nn_embedding/README.md (hunk 7 above)
Comment (LF-Vigneux) [reply papers/nn_embedding/README.md]: I forgot to remove that equation

**Review by CassNot — APPROVED** [review]
